### PR TITLE
redux of a change to use new engine APIs for Flutter.Frame events

### PR DESCRIPTION
**Pull request by devoncarew**
## Description

Redux of https://github.com/flutter/flutter/pull/34243

- update the `Flutter.Frame` event - sent over the VM service protocol - to use newly available engine apis. This will make the data match that reported via the Timeline events, and allows us to include timing info for both the dart and gpu portions of the frame

## Related Issues

none

## Tests

I added the following tests:

- validate that we fire a `Flutter.Frame` event in the binding tests
- update a devicelab test to fix an existing race condition - tickled by changes to when Flutter.Frame events are fired (`JSON-RPC error 103: Stream already subscribed`)

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 1]
```diff
@@ -198,6 +198,17 @@ mixin SchedulerBinding on BindingBase, ServicesBinding {
     window.onDrawFrame = _handleDrawFrame;
     SystemChannels.lifecycle.setMessageHandler(_handleLifecycleMessage);
     readInitialLifecycleStateFromNativeWindow();
+
+    if (!kReleaseMode) {
+      int frameNumber = 0;
+
+      window.onReportTimings = (List<FrameTiming> timings) {
```
Comment (liyuqian) [inline packages/flutter/lib/src/scheduler/binding.dart]: If the app has already set a `window.onReportTimings` listener, will this make it not work? Maybe we should also call the original `onReportTimings` if it's not null.

Comment (devoncarew) [reply packages/flutter/lib/src/scheduler/binding.dart]: This is called pretty early in app initialization; I don't know how likely it is that we'll lose the race condition with user code.

If a user of the API is responsible for delegating to the last person who set `onReportTimings`, how would that user know to stop calling the previous closure when that closure was no longer interested in events?

If we're concerned about this, I think an option here is to adjust the API to use a broadcast stream based approach (`Stream<FrameTiming> get onFrameTimings`). Clients could add themselves as a listener to the stream, and remove themselves, without having to worry about other potential users of the API interfering with them.

Comment (liyuqian) [reply packages/flutter/lib/src/scheduler/binding.dart]: Right, I think I'll keep the current API for a simple engine-to-framework communication, and add a broadcast stream for flexibility later as a pure framework pull request.

**Review by liyuqian — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/scheduler/binding.dart (hunk 1 above)
Comment (devoncarew) [reply packages/flutter/lib/src/scheduler/binding.dart]: OK, makes sense.

**Pull request merged**